File: engine/src/phase_manager.cpp
#include "phase_manager.h"
#include "logger.h"
namespace fe_engine {
	phase_manager::phase_manager() {
		this->m_current_phase = unit_affiliation::player;
	}
	unit_affiliation phase_manager::get_current_phase() const {
		return this->m_current_phase;
	}
	void phase_manager::cycle_phase(reference<map> m) {
		this->cycle_enum();
		for (size_t i = 0; i < m->get_unit_count(); i++) {
			reference<unit> u = m->get_unit(i);
			if (u->get_affiliation() == this->m_current_phase) {
				u->refresh_movement();
			}
		}
		this->log_phase();
	}
	void phase_manager::log_phase() {
		std::string phase;
		renderer::color color;
		switch (this->m_current_phase) {
		case unit_affiliation::player:
			phase = "Player";
			color = renderer::color::blue;
			break;
		case unit_affiliation::enemy:
			phase = "Enemy";
			color = renderer::color::red;
			break;
		case unit_affiliation::separate_enemy:
			phase = "Enemy";
			color = renderer::color::yellow;
			break;
		case unit_affiliation::ally:
			phase = "Ally";
			color = renderer::color::green;
			break;
		default:
			return;
			break;
		}
		logger::print(phase + " phase", color);
	}
	void phase_manager::cycle_enum() {
		this->m_current_phase = (unit_affiliation)(((int)this->m_current_phase + 1) % 4);
	}
}

File: rewrite/src/FEEngine/PhaseManager.cs
using System.Collections.Generic;

namespace FEEngine
{
    public class PhaseManager
    {
        public Unit.UnitAffiliation CurrentPhase { get; private set; }
        public PhaseManager(Unit.UnitAffiliation initialPhase = Unit.UnitAffiliation.Player)
        {
            CurrentPhase = initialPhase;
        }
        public void CyclePhase(Map map)
        {
            // i apologize for the spaghetti code
            switch (CurrentPhase)
            {
                case Unit.UnitAffiliation.Player:
                    CurrentPhase = Unit.UnitAffiliation.Enemy;
                    break;
                case Unit.UnitAffiliation.Enemy:
                    CurrentPhase = Unit.UnitAffiliation.ThirdEnemy;
                    break;
                case Unit.UnitAffiliation.ThirdEnemy:
                    CurrentPhase = Unit.UnitAffiliation.Ally;
                    break;
                case Unit.UnitAffiliation.Ally:
                    CurrentPhase = Unit.UnitAffiliation.Player;
                    break;
            }
            List<Unit> units = map.GetAllUnitsOfAffiliation(CurrentPhase);
            foreach (Unit unit in units)
            {
                unit.RefreshMovement();
            }
        }
    }
}
